File: backend/app/schemas/reports.py
"""Схемы отчётов."""

from uuid import UUID

from pydantic import BaseModel, Field


class StatusCount(BaseModel):
    status: str
    count: int


class ProjectReportSlice(BaseModel):
    project_id: UUID
    project_name: str
    active_total: int
    archived_total: int


class ReportsSummaryRead(BaseModel):
    project_id: UUID | None = None
    active_total: int = 0
    archived_total: int = 0
    closed_last_7_days: int = 0
    closed_last_30_days: int = 0
    with_due: int = 0
    overdue: int = 0
    due_soon: int = 0
    by_status: list[StatusCount] = Field(default_factory=list)
    by_project: list[ProjectReportSlice] = Field(default_factory=list)


class BurndownPoint(BaseModel):
    date: str
    closed_count: int
    remaining_active: int


class AssigneeSlice(BaseModel):
    assignee_id: UUID | None
    assignee_email: str | None
    overdue_count: int = 0
    in_progress_count: int = 0


class ReportsDashboardRead(BaseModel):
    project_id: UUID | None = None
    burndown: list[BurndownPoint] = Field(default_factory=list)
    in_progress: list[StatusCount] = Field(default_factory=list)
    overdue_by_assignee: list[AssigneeSlice] = Field(default_factory=list)
    time_total_seconds: int = 0

File: frontend/src/api/reports.ts
import { API_BASE_URL } from '../config'

import { apiJson } from './client'

export type ReportsSummary = {
  project_id: string | null
  active_total: number
  archived_total: number
  closed_last_7_days: number
  closed_last_30_days: number
  with_due: number
  overdue: number
  due_soon: number
  by_status: Array<{ status: string; count: number }>
  by_project: Array<{
    project_id: string
    project_name: string
    active_total: number
    archived_total: number
  }>
}

export type ReportsDashboard = {
  project_id: string | null
  burndown: Array<{ date: string; closed_count: number; remaining_active: number }>
  in_progress: Array<{ status: string; count: number }>
  overdue_by_assignee: Array<{
    assignee_id: string | null
    assignee_email: string | null
    overdue_count: number
    in_progress_count: number
  }>
  time_total_seconds: number
}

export function fetchReportsSummary(token: string, projectId?: string): Promise<ReportsSummary> {
  const q = projectId ? `?project_id=${encodeURIComponent(projectId)}` : ''
  return apiJson<ReportsSummary>(`/reports/summary${q}`, token)
}

export function fetchReportsDashboard(token: string, projectId?: string, days = 14): Promise<ReportsDashboard> {
  const q = new URLSearchParams({ days: String(days) })
  if (projectId) q.set('project_id', projectId)
  return apiJson<ReportsDashboard>(`/reports/dashboard?${q}`, token)
}

export async function downloadReportsExport(
  token: string,
  format: 'csv' | 'pdf',
  projectId?: string,
): Promise<void> {
  const q = new URLSearchParams({ format })
  if (projectId) q.set('project_id', projectId)
  const res = await fetch(`${API_BASE_URL}/reports/export?${q}`, {
    headers: { Authorization: `Bearer ${token}` },
  })
  if (!res.ok) throw new Error(`Export failed: ${res.status}`)
  const blob = await res.blob()
  const url = URL.createObjectURL(blob)
  const a = document.createElement('a')
  a.href = url
  a.download = format === 'pdf' ? 'near-report.pdf' : 'near-report.csv'
  a.click()
  URL.revokeObjectURL(url)
}

export function formatDuration(seconds: number): string {
  const h = Math.floor(seconds / 3600)
  const m = Math.floor((seconds % 3600) / 60)
  if (h > 0) return `${h} ч ${m} мин`
  return `${m} мин`
}
